src/lib/Hydra/View/NixLog.pm

```diff
@@ -5,32 +5,55 @@ use warnings;
 use base qw/Catalyst::View/;
 use Hydra::Helper::CatalystUtils;
 
+sub tail {
+    my ($filehandle, $n) = @_;
+    my @lines;
+    my $line_count = 0;
+
+    while (my $line = <$filehandle>) {
+        $lines[$line_count % $n] = $line;
+        $line_count++;
+    }
+
+    my $start = $line_count > $n ? $line_count % $n : 0;
+    my $end = $line_count > $n ? $n : $line_count;
+
+    my $result = "";
+    for my $i (0 .. $end - 1) {
+        $result .= $lines[($start + $i) % $n];
+    }
+    return $result;
+}
+
 sub process {
     my ($self, $c) = @_;
 
     my $logPath = $c->stash->{logPath};
 
     $c->response->content_type('text/plain; charset=utf-8');
 
-    my $fh = IO::Handle->new();
+    my $logFh = IO::Handle->new();
 
-    my $tail = int($c->stash->{tail} // "0");
+    my $tailLines = int($c->stash->{tail} // "0");
 
-    if ($logPath =~ /\.bz2$/) {
-        my $doTail = $tail ? "| tail -n '$tail'" : "";
-        open($fh, "-|", "bzip2 -dc < '$logPath' $doTail") or die;
+    if ($logPath =~ /\.zst$/) {
+        open($logFh, "-|", "zstd", "-dc", $logPath) or die;
```

Comment:
@SuperSandro2000 sorry, I got nerd snipped, when seeing shells being spawned with external inputs. Could you test this variant?